File: hexwar/lua/hexwar/render.lua
-- 렌더링 — SPEC §10
--
-- 프레임버퍼는 0..63999 인덱스의 정수 테이블이다. 파이썬의 bytearray 나
-- 타입스크립트의 Uint8Array 와 달리 루아 테이블은 값마다 태그가 붙은 64비트
-- 슬롯이라 메모리는 훨씬 크다. 대신 코드는 같다 — 도스식 '한 바이트 = 한
-- 픽셀' 모델이 언어와 무관하게 성립한다는 것을 보여 주는 자리다.

local F = require('hexwar.font')
local hexmap = require('hexwar.hexmap')
local PK = require('hexwar.picker')
local rngmod = require('hexwar.rng')
local units = require('hexwar.units')

local M = {}

M.SCR_W, M.SCR_H = 320, 200
M.VIEW = { 0, 0, 256, 168 }
M.PANEL = { 256, 0, 64, 200 }
M.MSG = { 0, 168, 256, 32 }

M.MAP_PX_W = hexmap.MAP_W * PK.HEX_W + PK.ODD_SHIFT
M.MAP_PX_H = hexmap.MAP_H * PK.ROW_STEP + (PK.HEX_H - PK.ROW_STEP)
M.CAM_MAX_X = math.max(0, M.MAP_PX_W - M.VIEW[3])
M.CAM_MAX_Y = math.max(0, M.MAP_PX_H - M.VIEW[4])

M.TILE_NAME = { [0] = 't_clear', 't_forest', 't_hill', 't_mountain',
                't_city', 't_river', 't_swamp', 't_sea' }
M.MINI_COLOR = { [0] = 24, 38, 56, 72, 120, 88, 104, 84 }

local function golden_dir()
  return os.getenv('HEXWAR_GOLDEN') or '../golden'
end

function M.load_palette(path)
  path = path or (golden_dir() .. '/palette.txt')
  local pal = {}
  local i = 0
  for line in io.lines(path) do
    local r, g, b = line:match('(%d+) (%d+) (%d+)')
    if r then
      pal[i] = { tonumber(r), tonumber(g), tonumber(b) }
      i = i + 1
    end
  end
  return pal
end

function M.load_sprites(path)
  path = path or (golden_dir() .. '/tiles.rle')
  local sprites = {}
  local name, w, h, buf, need, pos = nil, 0, 0, nil, 0, 0
  for line in io.lines(path) do
    if line ~= '' and line:sub(1, 1) ~= ';' then
      if need == 0 then
        local n, ws, hs = line:match('(%S+) (%d+) (%d+)')
        name, w, h = n, tonumber(ws), tonumber(hs)
        buf, need, pos = {}, w * h, 0
      else
        -- for 의 제어 변수는 상수이므로 지역 변수로 받아 쓴다(루아 5.4+)
        for cs, vs in line:gmatch('(%d+) (%d+)') do
          local cnt, val = tonumber(cs), tonumber(vs)
          for _ = 1, cnt do
            buf[pos] = val
            pos = pos + 1
          end
          need = need - cnt
        end
        if need == 0 then
          sprites[name] = { w = w, h = h, data = buf }
        end
      end
    end
  end
  return sprites
end

-- 프레임버퍼 -----------------------------------------------------------------
local FB = {}
FB.__index = FB
M.FB = FB

function M.new_fb(w, h)
  w = w or M.SCR_W
  h = h or M.SCR_H
  local d = {}
  for i = 0, w * h - 1 do d[i] = 0 end
  return setmetatable({ w = w, h = h, data = d }, FB)
end

function FB:clear(v)
  v = v or 0
  for i = 0, self.w * self.h - 1 do self.data[i] = v end
end

function FB:fill_rect(x, y, w, h, v)
  local x0 = math.max(0, x)
  local y0 = math.max(0, y)
  local x1 = math.min(self.w, x + w)
  local y1 = math.min(self.h, y + h)
  for yy = y0, y1 - 1 do
    local base = yy * self.w
    for xx = x0, x1 - 1 do self.data[base + xx] = v end
  end
end

function FB:frame_rect(x, y, w, h, v)
  self:fill_rect(x, y, w, 1, v)
  self:fill_rect(x, y + h - 1, w, 1, v)
  self:fill_rect(x, y, 1, h, v)
  self:fill_rect(x + w - 1, y, 1, h, v)
end

function FB:blit(sp, x, y, clip)
  local cx, cy, cw, ch = 0, 0, self.w, self.h
  if clip then cx, cy, cw, ch = clip[1], clip[2], clip[3], clip[4] end
  local x0 = math.max(x, cx)
  local y0 = math.max(y, cy)
  local x1 = math.min(x + sp.w, cx + cw)
  local y1 = math.min(y + sp.h, cy + ch)
  if x0 >= x1 or y0 >= y1 then return end
  local src, dst = sp.data, self.data
  for yy = y0, y1 - 1 do
    local srow = (yy - y) * sp.w - x
    local drow = yy * self.w
    for xx = x0, x1 - 1 do
      local v = src[srow + xx]
      if v ~= 0 then dst[drow + xx] = v end
    end
  end
end

function FB:text(s, x, y, color, clip)
  local cx, cy, cw, ch = 0, 0, self.w, self.h
  if clip then cx, cy, cw, ch = clip[1], clip[2], clip[3], clip[4] end
  for k = 1, #s do
    local gx = x + (k - 1) * F.ADV
    local rows = F.rows(s:sub(k, k))
    for ry = 0, F.FH - 1 do
      local bits = rows[ry]
      if bits ~= 0 then
        local py = y + ry
        if py >= cy and py < cy + ch then
          local base = py * self.w
          for bx = 0, F.FW - 1 do
            if (bits & (1 << (F.FW - 1 - bx))) ~= 0 then
              local px = gx + bx
              if px >= cx and px < cx + cw then self.data[base + px] = color end
            end
          end
        end
      end
    end
  end
end

function FB:to_ppm(pal)
  local lut = {}
  for i = 0, 255 do
    local c = pal[i] or { 0, 0, 0 }
    lut[i] = string.char(c[1] * 255 // 63, c[2] * 255 // 63, c[3] * 255 // 63)
  end
  local parts = { string.format('P6\n%d %d\n255\n', self.w, self.h) }
  local n = self.w * self.h
  local chunk, ci = {}, 0
  for i = 0, n - 1 do
    ci = ci + 1
    chunk[ci] = lut[self.data[i]]
    if ci == 4096 then
      parts[#parts + 1] = table.concat(chunk)
      chunk, ci = {}, 0
    end
  end
  if ci > 0 then parts[#parts + 1] = table.concat(chunk, '', 1, ci) end
  return table.concat(parts)
end

-- 렌더러 ---------------------------------------------------------------------
local Renderer = {}
Renderer.__index = Renderer
M.Renderer = Renderer

function M.new_renderer()
  return setmetatable({ fb = M.new_fb(), pal = M.load_palette(), sp = M.load_sprites() }, Renderer)
end

function Renderer:visible_rows(cam_y)
  local top = (cam_y - (PK.HEX_H - PK.ROW_STEP)) // PK.ROW_STEP
  local bot = (cam_y + M.VIEW[4]) // PK.ROW_STEP + 1
  return math.max(0, top), math.min(hexmap.MAP_H - 1, bot)
end

function Renderer:draw_roads(m, i, x, y)
  local nb, k = m:neighbors_with_dir(i)
  for j = 1, k do
    local d, ni = nb[j][1], nb[j][2]
    if (m.cells[ni] & 0x80) ~= 0 then
      self.fb:blit(self.sp['road' .. d], x, y, M.VIEW)
    end
  end
end

function Renderer:draw_unit(u, x, y)
  self.fb:blit(self.sp[string.format('u%d_%d', u.side, u.kind)], x + 8, y + 8, M.VIEW)
  local w = math.max(0, u.hp * 12 // 10)
  self.fb:fill_rect(x + 10, y + 25, 12, 2, 8)
  self.fb:fill_rect(x + 10, y + 25, w, 2, u.hp > 5 and 10 or 12)
end

function Renderer:draw_map(g, ui)
  local fb, m = self.fb, g.map
  local camx, camy = ui.cam_x, ui.cam_y
  fb:fill_rect(M.VIEW[1], M.VIEW[2], M.VIEW[3], M.VIEW[4], 0)
  local r0, r1 = self:visible_rows(camy)
  for row = r0, r1 do
    for col = 0, hexmap.MAP_W - 1 do
      local ox, oy = PK.hex_origin(col, row)
      local x, y = ox - camx, oy - camy
      if not (x <= -PK.HEX_W or x >= M.VIEW[3] or y <= -PK.HEX_H or y >= M.VIEW[4]) then
        local i = row * hexmap.MAP_W + col
        local fog = m.fog[i]
        if fog == hexmap.FOG_HIDDEN then
          fb:blit(self.sp.ov_black, x, y, M.VIEW)
        else
          local t = m.cells[i] & hexmap.TERRAIN_MASK
          fb:blit(self.sp[M.TILE_NAME[t]], x, y, M.VIEW)
          if (m.cells[i] & 0x80) ~= 0 then self:draw_roads(m, i, x, y) end
          if fog == hexmap.FOG_EXPLORED then
            fb:blit(self.sp.ov_dim, x, y, M.VIEW)
          else
            if ui.objective_idx[i] then fb:blit(self.sp.ov_obj, x, y, M.VIEW) end
            if ui.move_overlay[i] then fb:blit(self.sp.ov_move, x, y, M.VIEW) end
            if ui.attack_overlay[i] then fb:blit(self.sp.ov_attack, x, y, M.VIEW) end
            local uid = m.occupant[i]
            if uid ~= units.NO_UNIT then
              local u = g.pool:get(uid)
              if u then self:draw_unit(u, x, y) end
            end
            if ui.sel_idx == i then fb:blit(self.sp.ov_sel, x, y, M.VIEW) end
            if ui.cursor_idx == i then fb:blit(self.sp.ov_cursor, x, y, M.VIEW) end
          end
        end
      end
    end
  end
end

function Renderer:draw_minimap(g, ui, x, y)
  local m = g.map
  for row = 0, hexmap.MAP_H - 1 do
    for col = 0, hexmap.MAP_W - 1 do
      local i = row * hexmap.MAP_W + col
      local v
      if m.fog[i] == hexmap.FOG_HIDDEN then
        v = 0
      else
        v = M.MINI_COLOR[m.cells[i] & hexmap.TERRAIN_MASK]
        local uid = m.occupant[i]
        if uid ~= units.NO_UNIT and m.fog[i] ~= hexmap.FOG_EXPLORED then
          local u = g.pool:get(uid)
          if u then v = (u.side == 1) and 155 or 170 end
        end
      end
      self.fb:fill_rect(x + col * 2, y + row * 2, 2, 2, v)
    end
  end
  local vx = x + ui.cam_x * 2 // PK.HEX_W
  local vy = y + ui.cam_y * 2 // PK.ROW_STEP
  self.fb:frame_rect(vx, vy, M.VIEW[3] * 2 // PK.HEX_W, M.VIEW[4] * 2 // PK.ROW_STEP, 15)
end

function Renderer:draw_panel(g, ui)
  local fb = self.fb
  local px, py, pw, ph = M.PANEL[1], M.PANEL[2], M.PANEL[3], M.PANEL[4]
  fb:fill_rect(px, py, pw, ph, 8)
  fb:frame_rect(px, py, pw, ph, 7)
  fb:text('TURN ' .. g.turn, px + 4, py + 4, 15)
  fb:text('SIDE ' .. g.side, px + 4, py + 13, 15)
  self:draw_minimap(g, ui, px + 6, py + 24)
  local u = ui.sel_unit >= 0 and g.pool:get(ui.sel_unit) or nil
  local ty = py + 70
  if u then
    fb:text(units.KINDS[u.kind][1], px + 4, ty, 14)
    fb:text('HP ' .. u.hp, px + 4, ty + 10, 15)
    fb:text('MP ' .. u.mp, px + 4, ty + 19, 15)
    fb:text('AM ' .. u.ammo, px + 4, ty + 28, 15)
    fb:text('EN ' .. u.ent, px + 4, ty + 37, 15)
  else
    fb:text('NO UNIT', px + 4, ty, 7)
  end
  fb:text(ui:state_name(), px + 4, py + 136, 11)
end

function Renderer:draw_msg(g, ui)
  local fb = self.fb
  local mx, my, mw, mh = M.MSG[1], M.MSG[2], M.MSG[3], M.MSG[4]
  fb:fill_rect(mx, my, mw, mh, 0)
  fb:frame_rect(mx, my, mw, mh, 7)
  local lines = ui:ascii_log(g, 3)
  for i = 1, #lines do
    fb:text(lines[i]:sub(1, 41), mx + 3, my + 4 + (i - 1) * 9, (i == 1) and 15 or 7, M.MSG)
  end
end

function Renderer:draw_widgets(ui)
  local uimod = require('hexwar.ui')
  local queue = { ui.root }
  local qi = 1
  while qi <= #queue do
    local w = queue[qi]
    qi = qi + 1
    if w.visible then
      if w.kind == uimod.BUTTON then
        local on = w.enabled
        self.fb:fill_rect(w.x, w.y, w.w, w.h, on and 7 or 8)
        self.fb:frame_rect(w.x, w.y, w.w, w.h, on and 15 or 7)
        self.fb:text(w.label, w.x + 3, w.y + 3, on and 0 or 8)
      elseif w.kind == uimod.DIALOG then
        self.fb:fill_rect(w.x + 4, w.y + 4, w.w, w.h, 0)
        self.fb:fill_rect(w.x, w.y, w.w, w.h, 8)
        self.fb:frame_rect(w.x, w.y, w.w, w.h, 15)
        self.fb:frame_rect(w.x + 2, w.y + 2, w.w - 4, w.h - 4, 7)
        self.fb:text(w.label, w.x + 12, w.y + 14, 15)
      end
      for _, c in ipairs(w.children) do queue[#queue + 1] = c end
    end
  end
end

function Renderer:draw(g, ui)
  self:draw_map(g, ui)
  self:draw_panel(g, ui)
  self:draw_msg(g, ui)
  self:draw_widgets(ui)
  return self.fb
end

function Renderer:frame_hash()
  return rngmod.fnv1a(self.fb:to_ppm(self.pal))
end

return M

File: hexwar/py/hexwar/render.py
# -*- coding: utf-8 -*-
"""렌더링 — SPEC §10.

   화면은 320x200 8비트 인덱스 버퍼 하나다(모드 13h). 픽셀 하나가 팔레트
   번호 한 바이트라, 색을 바꾸는 것과 픽셀을 바꾸는 것이 분리된다 — 도스
   게임의 페이드인·팀 컬러·물결 애니메이션이 전부 팔레트만 건드려서 나온
   효과다.

   그리는 순서는 화가 알고리즘 그대로다. 헥스가 세로로 8픽셀씩 겹치므로
   위 행부터 그리면 아래 행이 자연스럽게 위를 덮는다.
"""

import io
import os

from . import font as F
from . import hexcoord as H
from . import picker as PK
from .hexmap import (FOG_HIDDEN, FOG_EXPLORED, TERRAIN_MASK,
                     T_MOVE, MAP_W, MAP_H)
from .rng import fnv1a
from .units import K_CHAR, NO_UNIT

SCR_W, SCR_H = 320, 200
VIEW = (0, 0, 256, 168)          # 맵이 보이는 창
PANEL = (256, 0, 64, 200)        # 오른쪽 정보 패널
MSG = (0, 168, 256, 32)          # 아래 메시지 줄

MAP_PX_W = MAP_W * PK.HEX_W + PK.ODD_SHIFT
MAP_PX_H = MAP_H * PK.ROW_STEP + (PK.HEX_H - PK.ROW_STEP)
CAM_MAX_X = max(0, MAP_PX_W - VIEW[2])
CAM_MAX_Y = max(0, MAP_PX_H - VIEW[3])

GOLDEN = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(
    os.path.abspath(__file__)))), 'golden')


class Sprite(object):
    __slots__ = ('w', 'h', 'data')

    def __init__(self, w, h, data):
        self.w = w
        self.h = h
        self.data = data


def load_palette(path=None):
    path = path or os.path.join(GOLDEN, 'palette.txt')
    pal = []
    for line in io.open(path, encoding='utf-8'):
        line = line.strip()
        if line:
            pal.append(tuple(int(x) for x in line.split()))
    return pal


def load_sprites(path=None):
    """RLE 코퍼스를 푼다. SPEC §10.2 — 이름 w h 다음에 (개수, 값) 쌍."""
    path = path or os.path.join(GOLDEN, 'tiles.rle')
    sprites = {}
    name = None
    w = h = 0
    buf = None
    need = 0
    for raw in io.open(path, encoding='utf-8'):
        line = raw.strip()
        if not line or line.startswith(';'):
            continue
        if need == 0:
            parts = line.split()
            name, w, h = parts[0], int(parts[1]), int(parts[2])
            buf = bytearray()
            need = w * h
            continue
        nums = [int(x) for x in line.split()]
        for i in range(0, len(nums), 2):
            c, v = nums[i], nums[i + 1]
            buf.extend(bytes([v]) * c)
            need -= c
        if need == 0:
            sprites[name] = Sprite(w, h, bytes(buf))
    return sprites


class Framebuffer(object):
    """한 장의 화면. 도스라면 A000:0000 이 가리키던 64KB 그 자체다."""

    __slots__ = ('w', 'h', 'data')

    def __init__(self, w=SCR_W, h=SCR_H):
        self.w = w
        self.h = h
        self.data = bytearray(w * h)

    def clear(self, v=0):
        for i in range(len(self.data)):
            self.data[i] = v

    def fill_rect(self, x, y, w, h, v):
        x0, y0 = max(0, x), max(0, y)
        x1, y1 = min(self.w, x + w), min(self.h, y + h)
        row = bytes([v]) * max(0, x1 - x0)
        for yy in range(y0, y1):
            self.data[yy * self.w + x0:yy * self.w + x1] = row

    def frame_rect(self, x, y, w, h, v):
        self.fill_rect(x, y, w, 1, v)
        self.fill_rect(x, y + h - 1, w, 1, v)
        self.fill_rect(x, y, 1, h, v)
        self.fill_rect(x + w - 1, y, 1, h, v)

    def blit(self, sp, x, y, clip=None):
        """인덱스 0을 건너뛰는 마스크 블릿. 클리핑은 원본 좌표를 잘라서 한다 —
           픽셀마다 범위 검사를 하면 8086 에서는 그 검사가 그리기보다 비쌌다."""
        cx, cy, cw, ch = clip if clip else (0, 0, self.w, self.h)
        x0 = max(x, cx)
        y0 = max(y, cy)
        x1 = min(x + sp.w, cx + cw)
        y1 = min(y + sp.h, cy + ch)
        if x0 >= x1 or y0 >= y1:
            return
        data = self.data
        src = sp.data
        for yy in range(y0, y1):
            srow = (yy - y) * sp.w - x
            drow = yy * self.w
            for xx in range(x0, x1):
                v = src[srow + xx]
                if v:
                    data[drow + xx] = v

    def text(self, s, x, y, color, clip=None):
        cx, cy, cw, ch = clip if clip else (0, 0, self.w, self.h)
        for ch_i, c in enumerate(s):
            gx = x + ch_i * F.ADV
            rows = F.rows(c)
            for ry in range(F.FH):
                bits = rows[ry]
                if not bits:
                    continue
                py = y + ry
                if py < cy or py >= cy + ch:
                    continue
                base = py * self.w
                for bx in range(F.FW):
                    if bits & (1 << (F.FW - 1 - bx)):
                        px = gx + bx
                        if cx <= px < cx + cw:
                            self.data[base + px] = color

    def to_ppm(self, palette):
        """P6 이진 PPM. 6비트 DAC 값을 8비트로 편다(v*255//63)."""
        head = ('P6\n%d %d\n255\n' % (self.w, self.h)).encode('ascii')
        out = bytearray(head)
        lut = bytearray()
        for (r, g, b) in palette:
            lut.extend((r * 255 // 63, g * 255 // 63, b * 255 // 63))
        for v in self.data:
            out.extend(lut[v * 3:v * 3 + 3])
        return bytes(out)


class Dirty(object):
    """더티 사각형 목록. SPEC §10.3 — 합집합이 너무 커지면 하나로 합친다.

       도스에서 이 최적화가 결정적이었던 이유: 320x200 전체를 다시 그리면
       64,000 바이트를 옮겨야 하는데, ISA 버스의 VRAM 쓰기는 느려서 그것만으로
       프레임을 다 썼다. 바뀐 칸 몇 개만 다시 그리면 수백 바이트로 끝난다.
    """

    __slots__ = ('rects',)

    def __init__(self):
        self.rects = []

    def add(self, x, y, w, h):
        if w <= 0 or h <= 0:
            return
        nr = [x, y, x + w, y + h]
        for r in self.rects:
            ux0, uy0 = min(r[0], nr[0]), min(r[1], nr[1])
            ux1, uy1 = max(r[2], nr[2]), max(r[3], nr[3])
            ua = (ux1 - ux0) * (uy1 - uy0)
            a = (r[2] - r[0]) * (r[3] - r[1]) + (nr[2] - nr[0]) * (nr[3] - nr[1])
            if a * 2 > ua:          # 합쳐도 낭비가 절반 미만이면 합친다
                r[0], r[1], r[2], r[3] = ux0, uy0, ux1, uy1
                return
        self.rects.append(nr)

    def area(self):
        return sum((r[2] - r[0]) * (r[3] - r[1]) for r in self.rects)

    def clear(self):
        self.rects = []


class Renderer(object):
    """맵·패널·메시지 줄을 그린다. 상태는 전부 밖(Game/Ui)에서 받는다."""

    __slots__ = ('fb', 'pal', 'sp', 'dirty')

    def __init__(self):
        self.fb = Framebuffer()
        self.pal = load_palette()
        self.sp = load_sprites()
        self.dirty = Dirty()

    # ---------------------------------------------------------------- 맵
    def visible_rows(self, cam_y):
        """화면에 걸치는 행 범위. 헥스가 32 높이인데 행 간격이 24라
           위아래로 한 행씩 더 그려야 잘리지 않는다."""
        top = (cam_y - (PK.HEX_H - PK.ROW_STEP)) // PK.ROW_STEP
        bot = (cam_y + VIEW[3]) // PK.ROW_STEP + 1
        return (max(0, top), min(MAP_H - 1, bot))

    def draw_map(self, g, ui):
        fb = self.fb
        m = g.map
        camx, camy = ui.cam_x, ui.cam_y
        fb.fill_rect(VIEW[0], VIEW[1], VIEW[2], VIEW[3], 0)
        r0, r1 = self.visible_rows(camy)
        for row in range(r0, r1 + 1):
            for col in range(MAP_W):
                ox, oy = PK.hex_origin(col, row)
                x = ox - camx
                y = oy - camy
                if x <= -PK.HEX_W or x >= VIEW[2] or y <= -PK.HEX_H or y >= VIEW[3]:
                    continue
                i = row * MAP_W + col
                fog = m.fog[i]
                if fog == FOG_HIDDEN:
                    self.fb.blit(self.sp['ov_black'], x, y, VIEW)
                    continue
                t = m.cells[i] & TERRAIN_MASK
                fb.blit(self.sp[TILE_NAME[t]], x, y, VIEW)
                if m.cells[i] & 0x80:
                    self._draw_roads(m, i, col, row, x, y)
                if fog == FOG_EXPLORED:
                    fb.blit(self.sp['ov_dim'], x, y, VIEW)
                    continue
                if i in ui.objective_idx:
                    fb.blit(self.sp['ov_obj'], x, y, VIEW)
                if i in ui.move_overlay:
                    fb.blit(self.sp['ov_move'], x, y, VIEW)
                if i in ui.attack_overlay:
                    fb.blit(self.sp['ov_attack'], x, y, VIEW)
                uid = m.occupant[i]
                if uid != NO_UNIT:
                    u = g.pool.get(uid)
                    if u is not None:
                        self._draw_unit(u, x, y)
                if ui.sel_idx == i:
                    fb.blit(self.sp['ov_sel'], x, y, VIEW)
                if ui.cursor_idx == i:
                    fb.blit(self.sp['ov_cursor'], x, y, VIEW)

    def _draw_roads(self, m, i, col, row, x, y):
        """이웃도 도로일 때만 그 방향으로 연결선을 그린다 — 도로가 끊기지 않는다."""
        for d, ni in m.neighbors_with_dir(i):
            if m.cells[ni] & 0x80:
                self.fb.blit(self.sp['road%d' % d], x, y, VIEW)

    def _draw_unit(self, u, x, y):
        self.fb.blit(self.sp['u%d_%d' % (u.side, u.kind)], x + 8, y + 8, VIEW)
        # 체력 막대 — 유닛 아이콘 아래 12x2
        w = max(0, u.hp * 12 // 10)
        self.fb.fill_rect(x + 10, y + 25, 12, 2, 8)
        self.fb.fill_rect(x + 10, y + 25, w, 2, 10 if u.hp > 5 else 12)

    # -------------------------------------------------------------- 패널
    def draw_panel(self, g, ui):
        fb = self.fb
        px, py, pw, ph = PANEL
        fb.fill_rect(px, py, pw, ph, 8)
        fb.frame_rect(px, py, pw, ph, 7)
        fb.text('TURN %d' % g.turn, px + 4, py + 4, 15)
        fb.text('SIDE %d' % g.side, px + 4, py + 13, 15)
        self.draw_minimap(g, ui, px + 6, py + 24)
        u = g.pool.get(ui.sel_unit) if ui.sel_unit >= 0 else None
        ty = py + 70
        if u is not None:
            from .units import KINDS
            fb.text(KINDS[u.kind][0], px + 4, ty, 14)
            fb.text('HP %d' % u.hp, px + 4, ty + 10, 15)
            fb.text('MP %d' % u.mp, px + 4, ty + 19, 15)
            fb.text('AM %d' % u.ammo, px + 4, ty + 28, 15)
            fb.text('EN %d' % u.ent, px + 4, ty + 37, 15)
        else:
            fb.text('NO UNIT', px + 4, ty, 7)
        fb.text(ui.state_name(), px + 4, py + 136, 11)

    def draw_minimap(self, g, ui, x, y):
        """헥스 하나를 2x2 픽셀로. 도스 워게임의 미니맵이 정확히 이 크기였다."""
        m = g.map
        for row in range(MAP_H):
            for col in range(MAP_W):
                i = row * MAP_W + col
                if m.fog[i] == FOG_HIDDEN:
                    v = 0
                else:
                    t = m.cells[i] & TERRAIN_MASK
                    v = MINI_COLOR[t]
                    uid = m.occupant[i]
                    if uid != NO_UNIT and m.fog[i] != FOG_EXPLORED:
                        u = g.pool.get(uid)
                        if u is not None:
                            v = 155 if u.side == 1 else 170
                self.fb.fill_rect(x + col * 2, y + row * 2, 2, 2, v)
        # 현재 보이는 영역 테두리
        vx = x + ui.cam_x * 2 // PK.HEX_W
        vy = y + ui.cam_y * 2 // PK.ROW_STEP
        self.fb.frame_rect(vx, vy, VIEW[2] * 2 // PK.HEX_W, VIEW[3] * 2 // PK.ROW_STEP, 15)

    # ---------------------------------------------------------- 메시지 줄
    def draw_msg(self, g, ui):
        fb = self.fb
        mx, my, mw, mh = MSG
        fb.fill_rect(mx, my, mw, mh, 0)
        fb.frame_rect(mx, my, mw, mh, 7)
        lines = ui.ascii_log(g, 3)
        for i, line in enumerate(lines):
            fb.text(line[:41], mx + 3, my + 4 + i * 9, 7 if i else 15, MSG)

    # ------------------------------------------------------------ 위젯
    def draw_widgets(self, ui):
        """위젯 트리를 앞에서 뒤로 그린다 — 히트 테스트와 정확히 반대 순서다.
           같은 트리를 두 방향으로 훑는 것이 도스 GUI 의 기본 골격이었다."""
        from .ui import BUTTON, DIALOG
        stack = [ui.root]
        while stack:
            w = stack.pop(0)
            if not w.visible:
                continue
            if w.kind == BUTTON:
                on = w.enabled
                self.fb.fill_rect(w.x, w.y, w.w, w.h, 7 if on else 8)
                self.fb.frame_rect(w.x, w.y, w.w, w.h, 15 if on else 7)
                self.fb.text(w.label, w.x + 3, w.y + 3, 0 if on else 8)
            elif w.kind == DIALOG:
                self.fb.fill_rect(w.x + 4, w.y + 4, w.w, w.h, 0)      # 그림자
                self.fb.fill_rect(w.x, w.y, w.w, w.h, 8)
                self.fb.frame_rect(w.x, w.y, w.w, w.h, 15)
                self.fb.frame_rect(w.x + 2, w.y + 2, w.w - 4, w.h - 4, 7)
                self.fb.text(w.label, w.x + 12, w.y + 14, 15)
            stack.extend(w.children)

    def draw(self, g, ui):
        self.draw_map(g, ui)
        self.draw_panel(g, ui)
        self.draw_msg(g, ui)
        self.draw_widgets(ui)
        return self.fb

    def frame_hash(self):
        return fnv1a(self.fb.to_ppm(self.pal))


TILE_NAME = ('t_clear', 't_forest', 't_hill', 't_mountain',
             't_city', 't_river', 't_swamp', 't_sea')
MINI_COLOR = (24, 38, 56, 72, 120, 88, 104, 84)
